theseus/core/vectorizer.py

```diff
@@ -306,6 +309,66 @@ def _vectorize(self):
                 vectorized_cost_fn, cost_fn_wrappers, batch_size
             )
 
+    @staticmethod
+    def _vectorized_retract_optim_vars(
+        delta: torch.Tensor,
+        ordering: Iterable[Manifold],
+        ignore_mask: Optional[torch.Tensor] = None,
+        force_update: bool = False,
+    ):
+        # Each variable type gets mapped to a tuple with:
+        #   - the variable that will hold the vectorized data
+        #   - all the variables of that type that will be vectorized together
+        #   - the delta slices that go into the batch
+        var_info: Dict[
+            Type[Manifold], Tuple[Manifold, List[Manifold], List[torch.Tensor]]
+        ] = {}
+
+        start_idx = 0
+        batch_size = -1
+        # Create var info by looping variables in the given order
+        # All variables of the same type get grouped together, and their deltas saved
+        for var in ordering:
+            if batch_size == -1:
+                batch_size = var.shape[0]
+            else:
+                assert batch_size == var.shape[0]
+
+            delta_slice = delta[:, start_idx : start_idx + var.dof()]
+            start_idx += var.dof()
+            if var.__class__ not in var_info:
+                var_info[var.__class__] = (var.copy(), [], [])
+            var_info[var.__class__][1].append(var)
+            var_info[var.__class__][2].append(delta_slice)
+
+        for _, (vectorized_var, var_list, delta_list) in var_info.items():
+            n_vars, dof = len(var_list), vectorized_var.dof()
+
+            # Get the vectorized tensor that has the current variable data.
+            # The resulting shape is (N * b, M), b is batch size, N is the number of
+            # variables in the group, and M is the data shape for this class
+            vectorized_data = torch.cat([v.data for v in var_list], dim=0)
+            assert (
+                vectorized_data.shape
+                == (n_vars * batch_size,) + vectorized_data.shape[1:]
+            )
+            # delta_list is a list of N tensors of shape (b, d), where
+            # d is var.dof(). After vectorization, we get tensor of shape (N * b, d)
+            vectorized_delta = torch.cat(delta_list, dim=0)
+            assert vectorized_delta.shape == (n_vars * batch_size, dof)
+
+            # Retract the vectorized data, then redistribute to the variables
+            vectorized_var.update(vectorized_data)
+            new_var = vectorized_var.retract(vectorized_delta)
+            start_idx = 0
+            for var in var_list:
+                retracted_data_slice = new_var[start_idx : start_idx + batch_size, :]
+                if ignore_mask is None or force_update:
+                    var.update(retracted_data_slice)
+                else:
+                    var.update(retracted_data_slice, batch_ignore_mask=ignore_mask)
```

Comment:
How this update affects the vectorized cost function?

Reply:
Good question. I think this should work, since `ignore_mask` applies to all variables, and it indexes over the batch. So we can first compute everything, then apply the (same) mask for each variable. But I'll try to add a unit tests for this as well, just in case.

Reply:
Once the unit test is passed. I think we can merge this.